### PR TITLE
fix: always reject invalid requests

### LINKED ISSUE #1091: Running ‘prism proxy’ with –errors flag doesn’t enforce required header
**Describe the bug**
I’m running prism as a proxy for another service, and the OpenAPI spec for the service specifies that an x-authorization header is required.

When I submit the request to the proxy, I see that prism is showing a validation violation due to missing header. However, I’m getting the response from the “live” service itself, rather than the error about the missing header from prism proxy. I’m using the ‘–errors’ flag. I’ve also confirmed that other sorts of errors ARE reported.

**To Reproduce**
Requires Docker
1. Clone the following repository: [TylerNielsen/prism-repro-issue](https://github.com/TylerNielsen/prism-repro-issue). I don't have a running service that I can share to reproduce this issue, so this repro will allow a docker container that can be spun up as a target. Note that this repo also contains the example OpenApi.json file.
2. Build the docker image from the cloned repo: `docker build -t prism-repro-1212 -f Dockerfile .`
3. Run the mock service in a container: `docker run -it --rm -p 8080:8080 prism-repro-1212`
4. Start the prism proxy by running this command: `prism proxy openapi.json http://127.0.0.1:8080 --errors`
5. Run the following command: `curl -X POST -H "Content-Type: application/json" http://0.0.0.0:4010/v1/presigned-url -d "{\"test\": \"this is a test\"}"`
6. See that the response is returned from the "live" service, but prism is throwing an error about the missing "x-company-auth" header.

**Expected behavior**
Since the `--errors` flag is included, the prism proxy service should reject the incoming request and return an error like it does for missing required properties in the body.

**Environment (remove any that are not applicable):**
 - Library version: 3.3.3
 - OS: Windows 10
 - Browser: N/A

**Additional context**
If you run prism as a mock server, rather than as a proxy, I see that the request without the x-company-auth header is rejected.

### PR DESCRIPTION
**Pull request by XVincentX**
This PR has a good thing and a bad thing.

The good: closes #1091, makes a better use of typings here and there and it also removes a deprecation warning
The bad: it closes the bug in a way that's not cool (basically duplicating some efforts here and there) and shows that it is probably time to resume the rethinking of the validation process that started some time ago. You can find more details on #1135

### CODE REVIEW

File: packages/core/src/factory.ts [hunk 1]
```diff
@@ -41,34 +43,36 @@ export function factory<Resource, Input, Output, Config extends IPrismConfig>(
...
+  ): E.Either<Error, ResourceAndValidation> => {
+    const validations = compact([
+      config.checkSecurity ? components.validateSecurity({ resource, element: input }) : undefined,
+      config.validateRequest ? components.validateInput({ resource, element: input }) : undefined,
+    ]);
+
+    return pipe(
+      eitherSequence(validations),
+      E.fold<NonEmptyArray<IPrismDiagnostic>, unknown, IPrismDiagnostic[]>(identity, () => []),
+      validations => E.right({ resource, validations })
```
Comment (karol-maciaszek) [inline packages/core/src/factory.ts:55]: Why do we even bother in returning `Either` if it cannot return Left? (I am aware that it is not related to this PR)

File: test-harness/specs/proxy/proxy.valid-request.txt [hunk 2]
```diff
@@ -0,0 +1,23 @@
+====test====
+Sending an invalid request through the proxy server returns a validation error
```
[the remark this answers is not in the document]
Comment (XVincentX) [reply test-harness/specs/proxy/proxy.valid-request.txt]: a80a847

File: packages/http/src/forwarder/__tests__/index.spec.ts [hunk 3]
```diff
@@ -108,11 +118,29 @@ describe('forward', () => {
...
+          hopByHopHeaders.forEach(hopHeader => {
+            expect(r.headers?.[hopHeader]).toBeUndefined();
+          })
       );
     });
   });
+
+  describe('and there are input validation errors', () => {
+    it('will refuse to forward and return an error', () =>
+      assertResolvesLeft(
```
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/forwarder/__tests__/index.spec.ts:133]: Done

File: packages/core/src/factory.ts (hunk 1 above)
Comment (XVincentX) [reply packages/core/src/factory.ts:55]: Great question, this can effectively be an `Option` rather than an Either — the reason why it's in this weird situation it's because it will eventually be an Either.

Now that both Proxy and Mock halt the request if it's invalid, with some more refactor this will behave as a regular Either and return 422 when the request is invalid.

I'll write up more details in #1135

Comment (karol-maciaszek) [reply packages/core/src/factory.ts:55]: Sounds good 👍

File: packages/http/src/forwarder/__tests__/index.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (karol-maciaszek) [reply packages/http/src/forwarder/__tests__/index.spec.ts:133]: 👍

File: packages/http/src/router/__tests__/utils.ts [hunk 4]
```diff
@@ -17,29 +17,29 @@ export function randomArray<T>(itemGenerator: () => T, length = 1): T[] {
...
-interface IRandomPathOptions {
+type IRandomPathOptions = {
   pathFragments?: number;
   includeTemplates?: boolean;
   trailingSlash?: boolean;
   leadingSlash?: boolean;
-}
+};
+
+const defaultRandomPathOptions: NonNullable<IRandomPathOptions> = {
```
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http/src/router/__tests__/utils.ts:27]: Very good catch. It turns out I was using the wrong type, I need a `DeepNonNullable`.

49bea78

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## Fixed
 
+- Prism will now correctly refuse invalid requests even when used with the `proxy` command [#1101](https://github.com/stoplightio/prism/pull/1101)
```
Comment (philsturgeon) [inline CHANGELOG.md:13]: so long as `--errors` is enabled?
